### ISSUE TITLE
Crash on injecting Forza Horizon 4

### ISSUE THREAD

**Issue by Tomatosauce288** [Wunkolo/UWPDumper#11]
As soon as the injector injects the DLL into FH4, the game just crashes and with it the dumper.

**Comment by MisterScratch** [comment]
Yep came here to post this.  It succeeds at injecting and then the game crashes before it can do whatever remote thread IPC is.

**Comment by Wunkolo** [comment]
I don't have Forza 4 myself to really test much of this.
Have you tried using something like Process Explorer to suspend the process before injecting?

**Comment by d1083168** [comment]
FH4 Demo works for some reason

**Comment by BennettStaley** [comment]
It's not related to permissions issues AFAIK. It seems to be blocking CreateThread.

No special windows or SDK versions from what i can see. And the folder layout is exactly the same as other titles. Albeit without the scrambled file names.

I am a firm believer that is is an anti-injection countermeasure, as the dumper worked just fine on press release 1. Even a dummy dll that just launches a command prompt window [/Cybellum/DoubleAgent](https://github.com/Cybellum/DoubleAgent) causes a crash.

 Where would I be able to add the debug parameter?

**Comment by Wunkolo** [comment]
Usually when a program crashes you'd be presented with some way to pick between closing the program or attempting to debug it. Which would allow you to pick a debug program like visual studio or ida or something to see the exact memory location, instructions, thread, and callstack where the crash occured.
![](https://i.imgur.com/aMEJceF.png)

If the crash happens at the Injector side then you can see where it crashes on the injector end but if DoubleAgent doesn't even seem to be able to get in then what it is probably doing is blocking the addition of loading additional DLLs at all somehow despite DoubleAgent being able to inject very early in the process creation. It also is capable of blocking it even with a suspended process somehow.

What do you mean exactly that it blocks CreateThread by the way? Is this in the injected DLL or in the injector itself and how did you derive that it's this? The injector runs `CreateRemoteThread` which pretty much gets the target process to run `CreateThread`. The injected DLL itself also creates a new thread for the actual dumping that occurs.

**Comment by BennettStaley** [comment]
Right, after looking further my comment about createThread seems to be incorrect.

It's definitely something inside the UWP wrapper itself, rather than the process it spawns, as suspending it has zero effect. I also found it impossible to attach x64dbg to put my, albeit uneducated, debugging skills to the test. lol.

The app doesn't seem to be crashing either, since I would expect the app to throw the error you included above. Instead its terminating without throwing an error or prompting debug, as I've seen a lot of UWP apps do.

> but if DoubleAgent doesn't even seem to be able to get in then what it is probably doing is blocking the addition of loading additional DLLs at all somehow despite DoubleAgent being able to inject very early in the process creation. It also is capable of blocking it even with a suspended process somehow.

bingo. DoubleAgents failure was a surprise to me.

**Comment by Wunkolo** [comment]
I don't know if DoubleAgent handles this. But any dll loaded into a UWP process has to have the [S-1-15-2-1](https://github.com/Wunkolo/UWPDumper/blob/master/UWPInjector/source/main.cpp#L121) (`ALL APPLICATION PACKAGES`) access control too to be injected. Unless DoubleAgent already works with other UWP processes just fine and bypasses this requirement.
Debugging a UWP process externally via visual studio or IDAPro is generally pretty spotty since the virtualized UWP space doesn't play nice with being debugged from the native windows side on top of whatever anti-debug that Forza might have. Still surprising that even when suspended it crashes.

I've known [some people](https://twitter.com/TheWack0lian) that have tried to crack the EFS statically by snooping keys but haven't heard much from those efforts in a while to feel that it has been successful. Sometime this weekend I'll probably be able to get a hold of Forza running on my workstation to see exactly what's going on and do some refactors for UWPDumper in general such as by transporting all the file data through the IPC rather than using [standard file IO](https://github.com/Wunkolo/UWPDumper/blob/master/UWPDumper/main.cpp#L153) into the tempstate folder and such.

Until then though any more detailed diagnostic info on this would help.

**Comment by d1083168** [comment]
@Wunkolo Like I've already told you in my E-Mail I could provide access to FH4 if you like. But it seems like you've already found a different source. But the offer is still there

**Comment by Wunkolo** [comment]
Can confirm that either the latest iteration of UWP or Forza itself is totally blocking any call to `LoadLibrary`. Possibly UWP itself if it even blocks DoubleAgent but the fact that it works on the demo version of the game makes me think it's just Forza possibly tapping into some UWP feature that polices what DLLs are allowed to get loaded into the process.

I tried getting it to call `LoadLibrary` from [QueueUserAPC](https://docs.microsoft.com/en-us/windows/desktop/api/processthreadsapi/nf-processthreadsapi-queueuserapc) and it would still cause it to crash. Attaching a debugger also seems to throw an insane amount of access violations to just about any debugger I attach to it as well (visual studio, ida, windbg, etc) so I can't get any more details from the crash.

Maybe I'm approaching this naively as I haven't touched UWP stuff in a very long time but I could for sure use some aid on this. There is definitely some low-level library blocking going on.

**Comment by BCVisGod** [comment]
Any idea on when this will be getting updated to work with fh4

**Comment by CtrLCC** [comment]
I'm curious about that too.

**Comment by genesisrwd** [comment]
Also interested if this is getting updated to work with FH4. Any news??

**Comment by Wunkolo** [comment]
No updates as of now unfortunately

**Comment by Blenux** [comment]
Try doing it this way, 

build release version of UWP Dumper, and download Process Hacker, 

Start Forza Horizon 4, and get to the main menu, then open process hacker, find the game process and right click and suspend it.

use UWP Dumper has normal and it should dump the files, not sure 100% sure it works has I tried Sea of Thieves but that has a custom pak compression that uModel can't open, about to try with GOW4 though.

**Comment by makslevental** [comment]
@Blenux doesn't work. suspended the process and it still crashed

**Comment by CtrLCC** [comment]
Btw most recent uwp game which is CrackDown 3 surprisingly working with Dumper.

**Comment by Wunkolo** [comment]
Those of you still interested in ripping Forza should check [this](https://cs.rin.ru/forum/viewtopic.php?p=1900512#p1900512) out

**Comment by tunip3** [comment]
do you have a mirror for those files as openload is gone?

**Comment by tunip3** [comment]
> Those of you still interested in ripping Forza should check [this](https://cs.rin.ru/forum/viewtopic.php?p=1900512#p1900512) out

do you have a copy of those files as openload is gone

**Comment by tunip3** [comment]
> Try doing it this way,
> 
> build release version of UWP Dumper, and download Process Hacker,
> 
> Start Forza Horizon 4, and get to the main menu, then open process hacker, find the game process and right click and suspend it.
> 
> use UWP Dumper has normal and it should dump the files, not sure 100% sure it works has I tried Sea of Thieves but that has a custom pak compression that uModel can't open, about to try with GOW4 though.
> 
> Edit* ok works with GOW4.
> 
> https://i.imgur.com/2ELLZCq.png

this does work but the files returned still have efs

**Comment by umustbeloggedintododat** [comment]
Doesn't work, it still crashes :(

**Comment by umustbeloggedintododat** [comment]
bump

**Comment by CE1CECL** [comment]
Does anyone here still have a owned copy of Forza Horizon 4 (not demo)?

If so, can someone try one of these steps below (and report back which one works, if any) (NOTE: the steps all might look the similar/same, but each one may have different end-results):

A. 
	1. Boot Up a Already-Installed-on-Disk Linux Distro on your PC, such as Ubuntu 22.04 (with a compatable pass-throughable GPU, not sure yet if the error box will cause problems inside of a vGPU such as [qemu-vmvga](http://github.com/CE1CECL/qemu-vmvga))
	2. `git clone --single-branch --depth=1 http://github.com/Scrut1ny/Hypervisor-Phantom`
	3. `cd Hypervisor-Phantom/Hypervisor-Phantom/`
	4. `./Auto-Hypervisor.sh`
	5. Go through installing both option 3 & option 4 (you'll need both set up on your machine)
	6. Import the `CE1VMBOX.xml` into `virt-manager` and adjust the XML if needed to your GPU setup (remove vmvga video if needed)
	7. Make sure under `<qemu:arg value="-cpu"/>` (the line under `-cpu` with the long string) that `hypervisor` is set to `+` (`+hypervisor`, use CTRL + F to find `hypervisor`)
	8. Install Windows 10 22H2 from Microsoft
	9. Download Forza Horizion 4 (not demo) from the Microsoft Store, install Forza Horizion 4, and open it. ! MAKE SURE THIS DOESN'T CRASH AT ALL !
	10. Download the [latest UWPDumper from GitHub](http://github.com/Wunkolo/UWPDumper/releases/tag/latest)
	11. While Forza Horizon 4 is running, use UWPDumper to attempt to dump Forza Horizon 4. ! STILL, MAKE SURE THIS DOESN'T CRASH AT ALL !
	12. If you came this far with success, report success @ http://github.com/Wunkolo/UWPDumper/issues/11

B.
	1. Boot Up a Already-Installed-on-Disk Linux Distro on your PC, such as Ubuntu 22.04 (with a compatable pass-throughable GPU, not sure yet if the error box will cause problems inside of a vGPU such as [qemu-vmvga](http://github.com/CE1CECL/qemu-vmvga))
	2. `git clone --single-branch --depth=1 http://github.com/Scrut1ny/Hypervisor-Phantom`
	3. `cd Hypervisor-Phantom/Hypervisor-Phantom/`
	4. `./Auto-Hypervisor.sh`
	5. Go through installing both option 3 & option 4 (you'll need both set up on your machine)
	6. Import the `CE1VMBOX.xml` into `virt-manager` and adjust the XML if needed to your GPU setup (remove vmvga video if needed)
	7. Make sure under `<qemu:arg value="-cpu"/>` (the line under `-cpu` with the long string) that `hypervisor` is set to `-` (`-hypervisor`, use CTRL + F to find `hypervisor`)
	8. Patch your Kernel with http://github.com/WCharacter/RDTSC-KVM-Handler and make sure the `diff` is set low enough to not crash later on (for me, on Forza Horizon 4 Demo, `2.9 / 8 = 0.3625` works at minimum without crash) (2.9GHz == My Host's CPU base speed, as a example)
	9. Install Windows 10 22H2 from Microsoft
	10. Download Forza Horizion 4 (not demo) from the Microsoft Store, install Forza Horizion 4, and open it. ! MAKE SURE THIS DOESN'T CRASH AT ALL !
	11. Download the [latest UWPDumper from GitHub](http://github.com/Wunkolo/UWPDumper/releases/tag/latest)
	12. While Forza Horizon 4 is running, use UWPDumper to attempt to dump Forza Horizon 4. ! STILL, MAKE SURE THIS DOESN'T CRASH AT ALL !
	13. If you came this far with success, report success @ http://github.com/Wunkolo/UWPDumper/issues/11

C.
	1. Boot Up a Already-Installed-on-Disk Linux Distro on your PC, such as Ubuntu 22.04 (with a compatable pass-throughable GPU, not sure yet if the error box will cause problems inside of a vGPU such as [qemu-vmvga](http://github.com/CE1CECL/qemu-vmvga))
	2. `git clone --single-branch --depth=1 http://github.com/Scrut1ny/Hypervisor-Phantom`
	3. `cd Hypervisor-Phantom/Hypervisor-Phantom/`
	4. `./Auto-Hypervisor.sh`
	5. Go through installing both option 3 & option 4 (you'll need both set up on your machine)
	6. Import the `CE1VMBOX.xml` into `virt-manager` and adjust the XML if needed to your GPU setup (remove vmvga video if needed)
	7. Make sure under `<qemu:arg value="-cpu"/>` (the line under `-cpu` with the long string) that `hypervisor` is set to `+` (`+hypervisor`, use CTRL + F to find `hypervisor`)
	8. Patch your Kernel with http://github.com/WCharacter/RDTSC-KVM-Handler and make sure the `diff` is set low enough to not crash later on (for me, on Forza Horizon 4 Demo, `2.9 / 8 = 0.3625` works at minimum without crash) (2.9GHz == My Host's CPU base speed, as a example)
	9. Install Windows 10 22H2 from Microsoft
	10. Download Forza Horizion 4 (not demo) from the Microsoft Store, install Forza Horizion 4, and open it. ! MAKE SURE THIS DOESN'T CRASH AT ALL !
	11. Download the [latest UWPDumper from GitHub](http://github.com/Wunkolo/UWPDumper/releases/tag/latest)
	12. While Forza Horizon 4 is running, use UWPDumper to attempt to dump Forza Horizon 4. ! STILL, MAKE SURE THIS DOESN'T CRASH AT ALL !
	13. If you came this far with success, report success @ http://github.com/Wunkolo/UWPDumper/issues/11

> Why?

Because it turns out, Forza Horizon 4 (both the demo version and non-demo version have RDTSC timing checks, but the timing checks may be off while in a detected hypervisor (VM), which may allow UWPDumper to run successfully.)

CE1VMBOX.xml: http://ce1cecl.github.io/CE1VMBOX.xml
